LSM: Revert groove to use update + insert and vendor hasUniqueRepresentation

cb22 opened the pull request:
# Motivation
Whenever we update an object, we always have the old object around. Previously, we used to do a lookup to find the old one in any case, to give a simpler `upsert` function signature to the state machine - but this was resulting in duplicated work.

# Changes
Revert that, going back to having an explicit `update` and `insert` with the state machine choosing which is better for its needs. Additionally, switch to using `stdx.equal_bytes` in the Groove `update` hot path. This required vendoring a small fix for https://github.com/ziglang/zig/issues/17592.



# Benchmarks
These two things combined give a nice performance bump:

```sh
# on tmpfs
./scripts/benchmark.sh --print-batch-timings --transfer-count=10_000_000 --account-count=10 --id-order=sequential
```

Before:
```
1362 batches in 29.25 s
load offered = 1000000 tx/s
load accepted = 341867 tx/s
```

After:
```
1356 batches in 25.52 s
load offered = 1000000 tx/s
load accepted = 391846 tx/s
```

cb22 committed:
Vendor hasUniqueRepresentation to account for packed structs

Works around https://github.com/ziglang/zig/issues/17592.

cb22 committed:
LSM: Revert groove to use update + insert
Save a lookup by requiring the old object to be passed in to the Groove, and
only check if constants.verify is true.

cb22 commented on src/lsm/groove.zig:
```diff
@@ -886,37 +888,43 @@ pub fn GrooveType(
                 groove.objects.put(new);
+                groove.objects_cache.upsert(new);
+            } else {
+                // TODO: Should we assert we actually updated something...?
```
We never did in the past - is updating an object to an identical state something we would expect is able to happen?

jorangreef replied:
Great question. Thanks for flagging this.

If the state machine calls `update()` then it should be unilateral (and we assert under `verify` that there is a change). Rather than the Groove doing magic to compare, to double check that the update is necessary, the Groove should trust the state machine to already know that the update is necessary.

If we later have state machines that might call `update()` but not know if the update is necessary, then we can add an extra `update_if_different()`.

jorangreef replied:
This hoists the comparison into the explicit policy/responsibility of the state machine, and then if it later needs a method to cede this responsibility to the groove, then that is also explicit.

cb22 commented on src/stdx.zig:
```diff
@@ -520,3 +520,46 @@ pub fn fstatfs(fd: i32, statfs_buf: *StatFs) usize {
+
+        .Struct => |info| {
+            // Packed structs are always unique.
+            if (info.backing_integer != null) {
```
If a struct is packed, we don't need to recursively check it like the unpacked case below, since Zig ensures packed structs can only contain other packed structsa.

matklad replied:
Let's add a couple of tests here though, just to sanity-check things.

matklad replied:
Also, for a good measure, let's assert that the bitsize of all fields adds up to the sizeof the whole struct (that is, that we don't have padding _bits_).

I am thinking about cases like 

```zig
const S = packed struct {
    x: u24,
    y: u3,
};
```

I think the language may work both as "the spare bits after `y` are zeroed" and "the spare bits after `y` are arbitrary". I _think_ they are actually zeroed, but let's assert that we just don't have such weird structs altogether.

jorangreef commented on src/state_machine.zig:
```diff
@@ -1006,8 +1006,8 @@ pub fn StateMachineType(
 
-            self.forest.grooves.accounts.upsert(&dr_account_new);
-            self.forest.grooves.accounts.upsert(&cr_account_new);
+            self.forest.grooves.accounts.update(&dr_account_new, dr_account);
```
```suggestion
            self.forest.grooves.accounts.update(dr_account, &dr_account_new);
```

So that progress moves from left to right, from old to new.

jorangreef replied:
A risk to my suggestion here though is that if an `insert(new)` is swapped to an `update(old, new)` then the caller needs to swap arguments.

However, if the interface remains `update(new, old)` then there is the risk that the caller does `(old, new)`... thinking progress moves positively, from left to right.

cb22 replied:
Yeah indeed - my ordering was based off the first case case, which also imitates our `target` first style for copies... A third option could be to make it an explicit struct? 

```zig
self.forest.grooves.accounts.update(.{ .old = dr_account, .new = &dr_account_new });
```

From what I can see, we don't have any other `update` functions floating around to base the style off.

> Two parameters of the same type here makes me nervous 😅

~ DJ, 1 year ago, on my AOF PR :smile:

cb22 replied:
After a bit more discussion:
- `target` first style for copies gets `const` protection, which we can't take advantage of here.
- There's no real convention, because this `.update()` method doesn't update a parameter with an existing one, but rather updates the groove based off of both.
- There's no performance cost to using a struct (asm generated on godbolt was identical and double checked on the benchmark).

I've gone with a struct.

cb22 committed:
Groove: Remove stdx.equal_bytes from the hot path entirely

The expectation is that the state machine will call .update() when there is in
fact an update. We still check stdx.equal_bytes gated behind constants.verify.

We also double check key_from_value and that the update is not a tombstone,
too.

batiati commented:
Can we add `hasUniqueRepresentation` to `tidy.banned(...)`?

cb22 committed:
Groove: Make update() take a struct to avoid ambiguity around param order

* target first style for copies gets const protection, which we can't take
  advantage of here.
* There's no real convention, because this .update() method doesn't update
  a parameter with an existing one, but rather updates the groove based off
  of both.
* There's no performance cost to using a struct (asm generated on godbolt
  was identical and double checked on the benchmark).

cb22 committed:
Tidy: add std hasUniqueRepresentation to banned list

batiati commented on src/stdx.zig:
```diff
@@ -520,3 +520,46 @@ pub fn fstatfs(fd: i32, statfs_buf: *StatFs) usize {
+// TODO(Zig): https://github.com/ziglang/zig/issues/17592.
+/// True if every value of the type `T` has a unique bit pattern representing it.
+/// In other words, `T` has no unused bits and no padding.
+pub fn hasUniqueRepresentation(comptime T: type) bool {
```
When vendoring, we've been adopting our snake_case convention.
Arguable here, since we should move back to std as soon we update Zig again 🤔.

```suggestion
pub fn has_unique_representation(comptime T: type) bool {
```

cb22 replied:
>  Arguable here, since we should move back to std as soon we update Zig again 🤔.

It's a small enough change either way, so I'm happy to adopt it.

matklad commented on src/lsm/groove.zig:
```diff
@@ -886,37 +888,55 @@ pub fn GrooveType(
                 const new_index = Helper.derive_index(new);
 
                 // Only update the indexes that change.
                 if (!std.meta.eql(old_index, new_index)) {
```
I wonder if comparing just `values` (and ignoring timestamps) would make sense here, benchmark-wise?.

cb22 replied:
`derive_index` does return `IndexType(field_name)` which boils down to just the `value`. It's only `derive_value` that creates the `CompositeKey` with both. `std.meta.eql` should boil away to `==` in these cases too.

matklad commented on src/lsm/groove.zig:
```diff
@@ -886,37 +888,55 @@ pub fn GrooveType(
             }
 
+            groove.objects.put(new);
+            groove.objects_cache.upsert(new);
```
This makes my [my Rust eye](https://static.wikia.nocookie.net/p__/images/7/73/Mad-Eye_Moody%27s_%22mad_eye%22.jpeg/revision/latest?cb=20160215100736&path-prefix=protagonist) twitch a bit. Isn't this shared mutable aliasing? I _think_ the `old` object might be the object from the cache (`maybe(old == old_from_cache)` in the if above). If that's the case, I think we _might_ be clobbering old here, in which case the `inline for` below would get wrong results when accessing `old` fields?

cb22 replied:
We have an explicit guard against this above; since I've been bitten by this before here :smile::

> assert(new != old);

I think that's enough to prevent it from happening?

matklad replied:
No, I think that's different. The `new` is an object on state_machine stack:

https://github.com/tigerbeetle/tigerbeetle/blob/c367d1dfdba783b8eb171cdfec5d9de0188517c5/src/state_machine.zig#L870

the `old` is an object from the cache:

https://github.com/tigerbeetle/tigerbeetle/blob/c367d1dfdba783b8eb171cdfec5d9de0188517c5/src/state_machine.zig#L813

So `new` and `old`  are different. 

However, when we do `groove.objects_cache.upsert(new);` this might invalidate `old` still, if it was in the cache. That is, after this line I would expect `new.*` and `old.*` to be equal, because the `old` memory was overwritten when we upserted `new`.

Not 100% sure that this could happen, but I think it does?

cb22 replied:
Ahhh yes, that is the case. Good catch!

In the past, I took an explicit stack copy of `old` within the function which prevented that. We could move the object cache upsert to the end - there's no reason it needs to be done before the index trees - with an explicit comment explaining the above. That way we can get correct behaviour without a stack copy, but it still feels a bit fragile...

matklad replied:
I think that's ok, as long as we document that with a `maybe`. 

I wonder if we even can _require_ that this is the object from the cache? It _must_ be, if we are updating, right?

matklad replied:
I am also _reeeeeally_ wondering why CI is green here :thinking: in `release`, I can see compiler hoisting the loads to before the cache update, which would give the correct behavior. 

But in debug, I expect the overwrite to be reflected robustly. Could you also take a look what's up with tests passing here? Either my theory about the overwrite is wrong (good case), or we somehow don't test this case at all (which would be pretty scary).

cb22 replied:
> Could you also take a look what's up with tests passing here? Either my theory about the overwrite is wrong (good case), or we somehow don't test this case at all (which would be pretty scary).

As far as I can tell, we don't have any tests (unit or fuzz) that actually check the result of the secondary index trees, and in this case, if `old.* == new.*` it'll just skip updating the secondary index so there's nothing that's hit on the way down :grimacing:...

Also just noticed the forest fuzz (which I forgot to run / update in this PR) has also been broken for a few weeks. I need to look at proper infra for fuzzing ASAP, but maybe as a workaround for now we can just run each fuzz for a fixed time on CI?

We previously did run some more fuzzes in CI, but with a fixed seed for a single iteration.

cb22 replied:
> I wonder if we even can require that this is the object from the cache? It must be, if we are updating, right?

It could in theory be a stack copy, but we don't have an explicit use case for that yet so I've made the check `assert(old_from_cache == old);`.

sentientwaffle replied:
> Also just noticed the forest fuzz (which I forgot to run / update in this PR) has also been broken for a few weeks.

I think that is fixed as part of https://github.com/tigerbeetle/tigerbeetle/pull/1234, which merged this morning.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -886,37 +888,55 @@ pub fn GrooveType(
+            if (constants.verify) {
+                const old_from_cache = groove.objects_cache.get(@field(old, primary_field)).?;
+                assert(@field(old_from_cache, primary_field) == @field(old, primary_field));
+                assert(old_from_cache.timestamp == old.timestamp);
```
Can you assert that `stdx.bytes_equal(Object, old, old_from_cache);`?

cb22 replied:
I'm now aserting they're actually one and the same: https://github.com/tigerbeetle/tigerbeetle/pull/1239#discussion_r1369774962

cb22 committed:
LSM: Fuzz fix, review changes

matklad commented on src/lsm/groove.zig:
```diff
@@ -946,6 +947,13 @@ pub fn GrooveType(
+            // objects_cache. If we upsert first, old.* == new.* and no secondary indexes will
+            // be updated!
+            groove.objects_cache.upsert(new);
+            groove.objects.put(new);
```
For an extra tripple check, can we assert here that now old & new _contents_ is the same?

cb22 replied:
Ok - so that won't actually hold; if the old value comes from the stash of the cache_map, and we insert a new value, it'll go into the cache - so now `old.* != new.*`. 

The assertion at the start, `assert(old_from_cache == old);` will still hold because if it comes from the stash from the caller, it should still come from the stash inside.

matklad commented on src/stdx.zig:
```diff
@@ -540,26 +540,133 @@ pub fn hasUniqueRepresentation(comptime T: type) bool {
-                return true;
+            // Only consider packed structs unique if they are byte aligned.
+            if (info.backing_integer) |backing_integer| {
+                return @sizeOf(@Type(.{ .Struct = info })) * 8 == @bitSizeOf(backing_integer);
```
`@sizeOf(T)` would be simpler I think?

cb22 replied:
... yes, I haven't had my Monster yet. :rofl:

jorangreef replied:
I think it's time to go pull an espresso too... ☕

cb22 committed:
LSM: Review changes

matklad commented on src/lsm/groove.zig:
```diff
@@ -910,9 +910,7 @@ pub fn GrooveType(
             assert(new != old);
 
-            if (has_id) {
-                assert(old.id == new.id);
```
:+1: I actually do line one-line without braces form quite a bit more, when it works

jorangreef replied:
Yes, we don't need to use kernel style indented returns etc.
